cpp/src/reshape/interleave_columns.cu

```diff
@@ -0,0 +1,120 @@
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+namespace {
+
+struct interleave_columns_functor
+{
+    template <typename T, typename... Args>
+    std::enable_if_t<not cudf::is_fixed_width<T>(), std::unique_ptr<cudf::column>>
+    operator()(Args&&... args)
+    {
+        CUDF_FAIL("Only fixed-width types are supported in interleave_columns.");
+    }
+
+    template <typename T>
+    std::enable_if_t<cudf::is_fixed_width<T>(), std::unique_ptr<cudf::column>>
+    operator()(table_view const& input,
+               bool create_mask,
+               rmm::mr::device_memory_resource *mr,
+               cudaStream_t stream = 0)
+    {
+        auto arch_column = input.column(0);
+        auto output_size = input.num_columns() * input.num_rows();
+        auto output = allocate_like(arch_column, output_size, mask_allocation_policy::NEVER, mr, stream);
+        auto device_input = table_device_view::create(input);
+        auto device_output = mutable_column_device_view::create(*output);
+        auto index_begin = thrust::make_counting_iterator<size_type>(0);
+        auto index_end = thrust::make_counting_iterator<size_type>(output_size);
+
+        auto func_value = [input=*device_input, divisor=input.num_columns()]
+            __device__ (size_type idx) {
+                return input.column(idx % divisor).element<T>(idx / divisor);
+            };
+
+        if (not create_mask)
+        {
+            thrust::transform(rmm::exec_policy(stream)->on(stream),
+                              index_begin,
+                              index_end,
+                              device_output->data<T>(),
+                              func_value);
+
+            return output;
+        }
+
+        auto func_validity = [input=*device_input, divisor=input.num_columns()]
+            __device__ (size_type idx) {
+                return input.column(idx % divisor).is_valid(idx / divisor);
+            };
+
+        thrust::transform_if(rmm::exec_policy(stream)->on(stream),
+                             index_begin,
+                             index_end,
+                             device_output->data<T>(),
+                             func_value,
+                             func_validity);
+
+        rmm::device_buffer mask;
+        size_type null_count;
+
+        std::tie(mask, null_count) = valid_if(index_begin,
+                                              index_end,
+                                              func_validity,
+                                              stream,
+                                              mr);
+
+        output->set_null_mask(std::move(mask), null_count);
+
+        return output;
+    }
+};
+
+} // anonymous namespace
+} // namespace detail
+
+std::unique_ptr<column>
+interleave_columns(table_view const& input,
+                   rmm::mr::device_memory_resource *mr)
+{
+    CUDF_EXPECTS(input.num_columns() > 0, "input must have at least one column to determine dtype.");
```

Comment:
Should not be an exception. Empty input table --> empty output table.

Reply:
That would avoid the exception, but is that the desired API? This function is designed to return a single column, only. Does returning a table make sense for that? I'm not talking about performance or memory consumption. I'm wondering about usability/expectations. It seems strange that the consumer would have to check if a column exists. That's effectively using `table `as a `maybe`. Is that what we want?

Reply:
Right, as discussed, exception is the best option here since you need to determine an output column type.

Reply:
We discussed this and decided an exception is appropriate in this case. The alternatives were:
1) return a table as if it were a maybe (which cannot indicate why the operation did not produce a column).
2) force the consumer to explicitly provide an expected dtype.
4) allow the function to return a "non-typed" column.